onnx/shape_inference/implementation.cc

```diff
@@ -283,6 +295,14 @@ static void InferShapesImpl(
       throw;
```

Comment:
It is unclear which exceptions are thrown vs which ones are caught and reported at the end. 
Seems like you keep the throw statement here because if we reach this catch block then it means Type/Shape inference failed for this node and we do not want to continue right?

In the earlier cases the exceptions are caught and stored in inference_errors because there was not enough info available to infer shapes and types so instead of returning we decide to continue ?

Reply:
Sorry I missed this earlier. My opinion is that this should be very much like error-messages produced by a compiler. Ideally, we should report as many distinct errors as possible. So, even for checkShapesAndTypes, if it is possible for the checker to continue and check the remaining model, it would be great to do so. However, whatever we do, it is important to avoid the possibility of issues like null-dereferencing. (E.g., like some of the issues that Ashwini recently fixed.) This is one of the main reasons to terminate early (in case we are not sure whether the subsequent code will robustly handle missing types etc.).

Reply:
So, I would say that the first priority should be to make sure the code-base is robust (in terms of avoiding assumptions that an input-type is known when doing inference for a node). When we have that robustness, we can relax more of these early-termination errors and continue with checking the rest of the model (thus, this is a secondary priority).